lib/pbio/drv/usb/stm32_usbd/usbd_desc.c

```diff
@@ -231,17 +377,88 @@ uint8_t *USBD_VCP_ConfigStrDescriptor(USBD_SpeedTypeDef speed, uint16_t *length)
+    /* Prevent unused argument(s) compilation warning */
+    UNUSED(speed);
+
+    *length = sizeof(USBD_BOSDesc);
+    return (uint8_t *)USBD_BOSDesc;
+}
+
+USBD_DescriptorsTypeDef USBD_Pybricks_Desc = {
+    .GetDeviceDescriptor = USBD_Pybricks_DeviceDescriptor,
+    .GetLangIDStrDescriptor = USBD_Pybricks_LangIDStrDescriptor,
+    .GetManufacturerStrDescriptor = USBD_Pybricks_ManufacturerStrDescriptor,
+    .GetProductStrDescriptor = USBD_Pybricks_ProductStrDescriptor,
+    .GetSerialStrDescriptor = USBD_Pybricks_SerialStrDescriptor,
+    .GetConfigurationStrDescriptor = USBD_Pybricks_ConfigStrDescriptor,
+    .GetInterfaceStrDescriptor = USBD_Pybricks_InterfaceStrDescriptor,
+    .GetBOSDescriptor = USBD_Pybricks_BOSDescriptor,
 };
+
+void USBD_Pybricks_Desc_Init(void) {
+    // the same firmware runs on both SPIKE Prime and Robot Inventor so we need
+    // to dynamically set the PID based on the variant
+    #ifdef PBDRV_CONFIG_USB_STM32F4_HUB_VARIANT_ADDR
+    #define VARIANT (*(uint32_t *)PBDRV_CONFIG_USB_STM32F4_HUB_VARIANT_ADDR)
+    if (VARIANT == 0) {
+        USBD_DeviceDesc[10] = LOBYTE(PBDRV_CONFIG_USB_PID_0);
+        USBD_DeviceDesc[11] = HIBYTE(PBDRV_CONFIG_USB_PID_0);
+    } else if (VARIANT == 1) {
+        USBD_DeviceDesc[10] = LOBYTE(PBDRV_CONFIG_USB_PID_1);
+        USBD_DeviceDesc[11] = HIBYTE(PBDRV_CONFIG_USB_PID_1);
+    }
+    #endif
+
+    // REVISIT: we could probably move this to the static initializer above,
+    // but it would probably be a bit fragile and ugly
+
+    uint8_t *ptr = &USBD_BOSDesc[USB_SIZ_BOS_DESC_CONST];
+
+    /* Add firmware version */
+    *ptr++ = USB_SIZ_PLATFORM_HDR + sizeof(firmware_version) - 1;
+    *ptr++ = USB_DEVICE_CAPABITY_TYPE;
+    *ptr++ = 0x05;
+    *ptr++ = 0x00;
+
+    pbio_uuid128_le_copy(ptr, pbio_gatt_firmware_version_char_uuid_128);
+    ptr += USB_SIZ_UUID;
+
+    memcpy(ptr, firmware_version, sizeof(firmware_version) - 1);
+    ptr += sizeof(firmware_version) - 1;
+
+    /* Add software (protocol) version */
+    *ptr++ = USB_SIZ_PLATFORM_HDR + sizeof(software_version) - 1;
+    *ptr++ = USB_DEVICE_CAPABITY_TYPE;
+    *ptr++ = 0x05;
+    *ptr++ = 0x00;
+
+    pbio_uuid128_le_copy(ptr, pbio_gatt_software_version_char_uuid_128);
+    ptr += USB_SIZ_UUID;
+
+    memcpy(ptr, software_version, sizeof(software_version) - 1);
+    ptr += sizeof(software_version) - 1;
+
+    /* Add hub capabilities */
+    *ptr++ = USB_SIZ_PLATFORM_HDR + PBIO_PYBRICKS_HUB_CAPABILITIES_VALUE_SIZE;
+    *ptr++ = USB_DEVICE_CAPABITY_TYPE;
+    *ptr++ = 0x05;
+    *ptr++ = 0x00;
+
+    pbio_uuid128_le_copy(ptr, pbio_pybricks_hub_capabilities_char_uuid);
+    ptr += USB_SIZ_UUID;
+
+    pbio_pybricks_hub_capabilities(ptr,
+        USBD_PYBRICKS_MAX_PACKET_SIZE - USB_SIZ_UUID,
+        PBSYS_APP_HUB_FEATURE_FLAGS,
+        PBSYS_PROGRAM_LOAD_MAX_PROGRAM_SIZE);
+    ptr += PBIO_PYBRICKS_HUB_CAPABILITIES_VALUE_SIZE;
+
+    /* Update bNumDeviceCaps field in BOS Descriptor */
+    USBD_BOSDesc[4] += 3;
```

Comment:
I think we will want to add another entry here to get the hub name. 16-bit UUID is 2A00.